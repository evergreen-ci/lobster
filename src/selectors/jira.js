// @flow strict

import { createSelector } from 'reselect';
import type { ReduxState, Line, Bookmark } from '../models';

const getLogLines = (state) => state.log.lines;
const getBookmarks = (state) => state.logviewer.bookmarks;

const getJiraTemplate = createSelector(
  getLogLines,
  getBookmarks,
  function(lines: Line[], bookmarks: Bookmark[]) {
    if (bookmarks.length === 0 || lines.length === 0) {
      return '';
    }

    let text = '{noformat}\n';
    for (let i = 0; i < bookmarks.length; i++) {
      const curr = bookmarks[i].lineNumber;
      if (curr >= lines.length) {
        text += '{noformat}';
        return text;
      }

      text += lines[curr].text + '\n';
      if ((i !== (bookmarks.length - 1)) && (bookmarks[i + 1].lineNumber !== (curr + 1))) {
        text += '...\n';
      }
    }
    text += '{noformat}';
    return text;
  }
);

// because this function memoises based on its parameters, prevent the user
// from passing the components props into it. Ideally, all users are relying on
// only the redux store's state to derive this data
export default function(state: ReduxState) {
  return getJiraTemplate(state);
}
