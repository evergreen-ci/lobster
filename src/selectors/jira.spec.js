import jira from './jira';

test('selectors-jira', function() {
  const lines = [
    { lineNumber: 0, text: 'line 0' },
    { lineNumber: 1, text: 'line 1' },
    { lineNumber: 2, text: 'line 2' },
    { lineNumber: 3, text: 'line 3' },
    { lineNumber: 4, text: 'line 4' },
    { lineNumber: 5, text: 'line 5' }
  ];

  let state = {
    log: {
      lines: lines
    },
    logviewer: {
      bookmarks: []
    }
  };

  expect(jira(state)).toBe('');

  state = Object.assign({}, state);
  state.logviewer.bookmarks = [
    { lineNumber: 0 },
    { lineNumber: 3 },
    { lineNumber: 5 },
    { lineNumber: 7 }
  ];

  expect(jira(state)).toEqual(expect.stringContaining('line 5'));
  expect(jira(state)).not.toEqual(expect.stringContaining('line 1'));
  expect(jira(state)).not.toEqual(expect.stringContaining('line 6'));
  expect(jira(state)).not.toEqual(expect.stringContaining('line 7'));
  expect(jira(state)).toMatchSnapshot();
});
