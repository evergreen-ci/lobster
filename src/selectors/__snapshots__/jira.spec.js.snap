// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors-jira 1`] = `
"{noformat}
line 0
...
line 3
...
line 5
...
{noformat}"
`;
